File: OneToOneCallSample/app/src/main/java/com/stringee/apptoappcallsample/stringee/service/MediaProjectionCallService.java
package com.stringee.apptoappcallsample.stringee.service;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ServiceInfo;
import android.os.Build;
import android.os.IBinder;

import androidx.annotation.Nullable;
import androidx.core.app.ServiceCompat;

import com.stringee.apptoappcallsample.stringee.common.CallConstants;
import com.stringee.apptoappcallsample.stringee.common.CallNotificationManager;
import com.stringee.apptoappcallsample.stringee.common.CallUtils;
import com.stringee.apptoappcallsample.stringee.manager.CallSession;
import com.stringee.apptoappcallsample.stringee.manager.StringeeCallManager;

/** Foreground service that owns the MediaProjection lifecycle for Call2 screen sharing. */
public class MediaProjectionCallService extends Service {
    public static final String EXTRA_PERMISSION_DATA = "media_projection_permission_data";
    private boolean stoppingFromSession;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent == null || !CallConstants.ACTION_START_IN_CALL.equals(intent.getAction())) {
            stopSelf(startId);
            return START_NOT_STICKY;
        }
        try {
            int type = Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
                    ? ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION : 0;
            ServiceCompat.startForeground(this, CallConstants.MEDIA_NOTIFICATION_ID,
                    CallNotificationManager.getInstance(this).buildMediaProjection(), type);
        } catch (RuntimeException exception) {
            CallUtils.reportException(MediaProjectionCallService.class, exception);
            stopSelf(startId);
            return START_NOT_STICKY;
        }
        Intent permissionData;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            permissionData = intent.getParcelableExtra(EXTRA_PERMISSION_DATA, Intent.class);
        } else {
            permissionData = getLegacyPermissionData(intent);
        }
        CallSession session = StringeeCallManager.getInstance(this).getSession();
        if (session == null || permissionData == null) {
            stopService();
            return START_NOT_STICKY;
        }
        session.startScreenShare(this, permissionData);
        return START_NOT_STICKY;
    }

    @SuppressWarnings("deprecation")
    private Intent getLegacyPermissionData(Intent intent) {
        return intent.getParcelableExtra(EXTRA_PERMISSION_DATA);
    }

    public void stopService() {
        stoppingFromSession = true;
        ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_REMOVE);
        CallNotificationManager.getInstance(this).cancel(CallConstants.MEDIA_NOTIFICATION_ID);
        stopSelf();
    }

    @Override
    public void onDestroy() {
        ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_REMOVE);
        CallNotificationManager.getInstance(this).cancel(CallConstants.MEDIA_NOTIFICATION_ID);
        if (!stoppingFromSession) {
            CallSession session = StringeeCallManager.getInstance(this).getSession();
            if (session != null) {
                session.onMediaProjectionServiceDestroyed(this);
            }
        }
        super.onDestroy();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: Kotlin-OneToOneCallSample/app/src/main/java/com/stringee/kotlin_onetoonecallsample/stringee/service/MediaProjectionCallService.kt
package com.stringee.kotlin_onetoonecallsample.stringee.service

import android.app.Service
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.IBinder
import androidx.core.app.ServiceCompat
import com.stringee.kotlin_onetoonecallsample.stringee.common.CallConstants
import com.stringee.kotlin_onetoonecallsample.stringee.common.CallNotificationManager
import com.stringee.kotlin_onetoonecallsample.stringee.common.CallUtils
import com.stringee.kotlin_onetoonecallsample.stringee.manager.CallSession
import com.stringee.kotlin_onetoonecallsample.stringee.manager.StringeeCallManager

/** Foreground service that owns the MediaProjection lifecycle for Call2 screen sharing. */
class MediaProjectionCallService : Service() {
    private var stoppingFromSession = false

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent == null || CallConstants.ACTION_START_IN_CALL != intent.getAction()) {
            stopSelf(startId)
            return START_NOT_STICKY
        }
        try {
            val type = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
                ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION
            else
                0
            ServiceCompat.startForeground(
                this, CallConstants.MEDIA_NOTIFICATION_ID,
                CallNotificationManager.Companion.getInstance(this).buildMediaProjection(), type
            )
        } catch (exception: RuntimeException) {
            CallUtils.reportException(MediaProjectionCallService::class.java, exception)
            stopSelf(startId)
            return START_NOT_STICKY
        }
        val permissionData: Intent?
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            permissionData =
                intent.getParcelableExtra<Intent?>(EXTRA_PERMISSION_DATA, Intent::class.java)
        } else {
            permissionData = getLegacyPermissionData(intent)
        }
        val session: CallSession? = StringeeCallManager.Companion.getInstance(this).session
        if (session == null || permissionData == null) {
            stopService()
            return START_NOT_STICKY
        }
        session.startScreenShare(this, permissionData)
        return START_NOT_STICKY
    }

    @Suppress("deprecation")
    private fun getLegacyPermissionData(intent: Intent): Intent? {
        return intent.getParcelableExtra<Intent?>(EXTRA_PERMISSION_DATA)
    }

    fun stopService() {
        stoppingFromSession = true
        ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_REMOVE)
        CallNotificationManager.Companion.getInstance(this)
            .cancel(CallConstants.MEDIA_NOTIFICATION_ID)
        stopSelf()
    }

    override fun onDestroy() {
        ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_REMOVE)
        CallNotificationManager.Companion.getInstance(this)
            .cancel(CallConstants.MEDIA_NOTIFICATION_ID)
        if (!stoppingFromSession) {
            val session: CallSession? = StringeeCallManager.Companion.getInstance(this).session
            if (session != null) {
                session.onMediaProjectionServiceDestroyed(this)
            }
        }
        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    companion object {
        const val EXTRA_PERMISSION_DATA: String = "media_projection_permission_data"
    }
}
